c/tests/test_minimal_cpp.cpp

```diff
@@ -78,12 +80,163 @@ test_table_basics()
+     * so we have to use a functional method. This is a copy of the cmp
+     * from fwdpp.  Only difference is the final time comparison
+     * (fwdpp table times to forwards). */
+    const auto cmp = [&node_time](const _edge &lhs, const _edge &rhs) {
+        auto tl = node_time[lhs.parent];
+        auto tr = node_time[rhs.parent];
+        if (tl == tr) {
+            if (lhs.parent == rhs.parent) {
+                if (lhs.child == rhs.child) {
+                    return lhs.left < rhs.left;
+                }
+                return lhs.child < rhs.child;
+            }
+            return lhs.parent < rhs.parent;
+        }
+        return tl < tr;
+    };
+
+    assert(start == 0);
+    /* Let's not bother with metadata */
+    assert(edges->metadata_length == 0);
+
+    sorted_edges.reserve(num_edges);
+    for (j = 0; j < num_edges; j++) {
+        sorted_edges.emplace_back(
+            edges->left[j], edges->right[j], edges->parent[j], edges->child[j]);
+    }
+
+    std::sort(begin(sorted_edges), end(sorted_edges), cmp);
+
+    for (j = 0; j < num_edges; j++) {
+        edges->left[j] = sorted_edges[j].left;
+        edges->right[j] = sorted_edges[j].right;
+        edges->parent[j] = sorted_edges[j].parent;
+        edges->child[j] = sorted_edges[j].child;
+    }
+    return 0;
+}
+
+void
+test_edge_sorting()
+{
+    std::cout << "test_edge_sorting" << endl;
+    tsk_table_collection_t tables;
+    tsk_id_t n = 10;
+    tsk_id_t j;
+    tsk_id_t ret = tsk_table_collection_init(&tables, 0);
+    assert(ret == 0);
+
+    tables.sequence_length = 1.0;
+    /* Make a stick tree */
+    /* Add nodes and edges */
+    for (j = 0; j < n; j++) {
+        ret = tsk_node_table_add_row(
+            &tables.nodes, TSK_NODE_IS_SAMPLE, j + 1, TSK_NULL, TSK_NULL, NULL, 0);
+        assert(ret == j);
+    }
+    for (j = n - 1; j > 0; j--) {
+        tsk_edge_table_add_row(&tables.edges, 0, 1, j, j - 1, NULL, 0);
+    }
+    assert(tables.nodes.num_rows == (tsk_size_t) n);
+    assert(tables.edges.num_rows == (tsk_size_t) n - 1);
+
+    /* Make sure the edges are unsorted */
+    ret = tsk_table_collection_check_integrity(&tables, TSK_CHECK_EDGE_ORDERING);
+    assert(ret == TSK_ERR_EDGES_NOT_SORTED_PARENT_TIME);
+
+    /* Sort the tables */
+    tsk_table_sorter_t sorter;
+    ret = tsk_table_sorter_init(&sorter, &tables, 0);
+    assert(ret == 0);
+    /* Set the sort_edges to our local C++ version. We could also set some
+     * persistent state in sorter.params if we wanted to. */
+    sorter.sort_edges = cpp_sort_edges;
+    ret = tsk_table_sorter_run(&sorter, NULL);
+    assert(ret == 0);
+    tsk_table_sorter_free(&sorter);
+
+    /* Make sure the edges are now sorted */
+    ret = tsk_table_collection_check_integrity(&tables, TSK_CHECK_EDGE_ORDERING);
+    assert(ret == 0);
+
+    tsk_table_collection_free(&tables);
+}
+
+int
+sort_edges_raises_exception(tsk_table_sorter_t *sorter, tsk_size_t start)
```

Comment:
One thing to also test is throwing a non-exception type.  I seriously hope no one would ever do it, but it is allowed:

```cpp
{
throw 42; // Yep, this is valid :)
}

Reply:
Eek, that's weird. I can imagine all sorts of creative abuses of that one!

Tested now, thanks.